File: packages/cdk/programs/hue.py
import websocket
import json
import requests
import time
import logging
from logging import handlers
import threading
import os

# Set up logging
logs_dir = "/tmp/logs"
if not os.path.exists(logs_dir):
    os.makedirs(logs_dir)

log_filename = os.path.join(logs_dir, "hue.log")
logger = logging.getLogger("HueLogger")
logger.setLevel(logging.INFO)
handler = logging.handlers.TimedRotatingFileHandler(
    log_filename, when="H", interval=1, backupCount=7 * 24
)
formatter = logging.Formatter('%(asctime)s - %(message)s')
handler.setFormatter(formatter)
logger.addHandler(handler)

# Philips Hue Bridge IP and credentials
PHILIPS_HUE_BRIDGE_IP = "192.168.86.115"
PHILIPS_HUE_USER = "t3SH1fzd1zG6tY5hqygvxFprZkQs1FfFatJwoG-L"

# WebSocket URL (from your WebSocket API)
WS_URL = "wss://zt4cmsh5r8.execute-api.us-east-1.amazonaws.com/prod"

# Function to list lights from the Philips Hue Bridge
def list_lights():
    try:
        url = f"http://{PHILIPS_HUE_BRIDGE_IP}/api/{PHILIPS_HUE_USER}/lights"
        response = requests.get(url)
        if response.status_code == 200:
            logger.info("Successfully retrieved lights")
            return response.json()
        else:
            logger.error(f"Failed to retrieve lights: {response.status_code} - {response.content}")
            return {"error": "Failed to retrieve lights"}
    except Exception as e:
        logger.error(f"Error while fetching lights: {e}")
        return {"error": str(e)}

# Function to toggle (trigger) light on or off
def trigger_light(light_id, state):
    try:
        url = f"http://{PHILIPS_HUE_BRIDGE_IP}/api/{PHILIPS_HUE_USER}/lights/{light_id}/state"
        payload = {"on": state}
        response = requests.put(url, json=payload)
        if response.status_code == 200:
            logger.info(f"Successfully set light {light_id} to {'on' if state else 'off'}")
            return {"success": True, "light_id": light_id, "state": state}
        else:
            logger.error(f"Failed to set light {light_id}: {response.status_code} - {response.content}")
            return {"error": "Failed to set light state"}
    except Exception as e:
        logger.error(f"Error while triggering light: {e}")
        return {"error": str(e)}

# WebSocket event handlers
def on_message(ws, message):
    logger.info(f"Received message: {message}")
    try:
        data = json.loads(message)

        # Handle the "list_lights" command
        if data.get("command") == "list_lights":
            lights = list_lights()

            # Send the result back to the WebSocket
            ws.send(json.dumps({
                "response": "list_lights",
                "data": lights
            }))

        # Handle the "trigger_light" command
        elif data.get("command") == "trigger_light":
            # Access data fields under the "data" key
            light_data = data.get("data", {})
            light_id = light_data.get("lightId")
            state = light_data.get("state")

            if light_id is not None and state is not None:
                result = trigger_light(light_id, state)

                # Send the result back to the WebSocket
                ws.send(json.dumps({
                    "response": "trigger_light_response",
                    "data": result
                }))
            else:
                logger.error("Invalid lightId or state received")
                ws.send(json.dumps({
                    "response": "trigger_light_response",
                    "error": "Invalid lightId or state"
                }))
    except Exception as e:
        logger.error(f"Error processing message: {e}")

def on_open(ws):
    logger.info("WebSocket connection opened")

def on_error(ws, error):
    logger.error(f"WebSocket error: {error}")

def on_close(ws, close_status_code, close_msg):
    logger.info(f"WebSocket closed: {close_status_code} - {close_msg}")
    # No recursive call here

# Start the WebSocket client with reconnection logic
def run_websocket_client():
    while True:
        try:
            ws = websocket.WebSocketApp(
                WS_URL,
                on_open=on_open,
                on_message=on_message,
                on_error=on_error,
                on_close=on_close
            )
            ws.run_forever()
        except Exception as e:
            logger.error(f"Exception in WebSocket connection: {e}")
        logger.info("WebSocket connection closed, reconnecting in 5 seconds")
        time.sleep(5)

if __name__ == "__main__":
    logger.info("Starting WebSocket client for Philips Hue control")
    run_websocket_client()

File: packages/cdk/lambda/casa-integrations/providers/hue.ts
// How to deploy:
// scp programs/hue.py drabindr@casaguard:/home/drabindr/ && ssh drabindr@casaguard 'sudo reboot'

import WebSocket from 'ws';
import { promisify } from 'util';

const WS_URL = 'wss://zt4cmsh5r8.execute-api.us-east-1.amazonaws.com/prod'; // WebSocket URL

// Promisified delay function for waiting for the response
const wait = promisify(setTimeout);

// Function to trigger the list_lights command over the WebSocket
export const listLights = async () => {
  return new Promise((resolve, reject) => {
    const ws = new WebSocket(WS_URL);

    let isResponseReceived = false;

    // On WebSocket connection open, send the list_lights command
    ws.on('open', () => {
      console.log('Connected to WebSocket, sending list_lights command...');
      ws.send(JSON.stringify({ command: 'list_lights' }));
    });

    // Handle WebSocket messages (response from the WebSocket)
    ws.on('message', (message: string) => {
      console.log(`Received message from WebSocket: ${message}`);
      const parsedMessage = JSON.parse(message);

      // Ignore the first message if it's an empty list_lights response
      if (parsedMessage.data && Object.keys(parsedMessage.data).length === 0) {
        console.log('Ignoring initial empty list_lights response...');
        return;
      }

      // Check if we received the real lights data
      if (parsedMessage.data && Object.keys(parsedMessage.data).length > 0) {
        isResponseReceived = true;
        resolve(parsedMessage.data);  // Resolve with the lights data
        ws.close();
      }
    });

    // Handle WebSocket errors
    ws.on('error', (error: any) => {
      console.error(`WebSocket error: ${error}`);
      if (!isResponseReceived) {
        reject(new Error(`WebSocket error: ${error}`));
      }
      ws.close();
    });

    // Handle WebSocket close event
    ws.on('close', () => {
      console.log('WebSocket connection closed');
      if (!isResponseReceived) {
        reject(new Error('WebSocket closed before a response was received'));
      }
    });

    // Set a timeout in case no response is received in 10 seconds
    wait(10000).then(() => {
      if (!isResponseReceived) {
        reject(new Error('No response from WebSocket in time'));
        ws.close();
      }
    });
  });
};

// Function to trigger a light (on/off) over the WebSocket
export const triggerLight = async (lightId: string, state: boolean) => {
  return new Promise((resolve, reject) => {
    const ws = new WebSocket(WS_URL);

    let isResponseReceived = false;

    // On WebSocket connection open, send the trigger command
    ws.on('open', () => {
      console.log('Connected to WebSocket, sending trigger command...');
      ws.send(
        JSON.stringify({
          command: 'trigger_light',
          data: {
            lightId,
            state,
          },
        })
      );
    });

    // Handle WebSocket messages (response from the WebSocket)
    ws.on('message', (message: string) => {
      console.log(`Received message from WebSocket: ${message}`);
      const parsedMessage = JSON.parse(message);

      // Check if the response contains the light_id and success for the trigger command
      if (parsedMessage.data && parsedMessage.data.light_id === lightId) {
        const { success, state } = parsedMessage.data;

        if (success) {
          isResponseReceived = true;
          resolve(parsedMessage.data);  // Resolve with the result of the trigger command
          ws.close();
        } else {
          reject(new Error(`Failed to trigger light ${lightId}`));
          ws.close();
        }
      }
    });

    // Handle WebSocket errors
    ws.on('error', (error: any) => {
      console.error(`WebSocket error: ${error}`);
      if (!isResponseReceived) {
        reject(new Error(`WebSocket error: ${error}`));
      }
      ws.close();
    });

    // Handle WebSocket close event
    ws.on('close', () => {
      console.log('WebSocket connection closed');
      if (!isResponseReceived) {
        reject(new Error('WebSocket closed before a response was received'));
      }
    });

    // Set a timeout in case no response is received in 5 seconds
    setTimeout(() => {
      if (!isResponseReceived) {
        reject(new Error(`No response for light ${lightId} in time`));
        ws.close();
      }
    }, 5000); // Timeout after 5 seconds
  });
};